hubspot/plugin.py

```diff
@@ -454,57 +454,95 @@ class HubSpotSearchCondition:
         "LTE",
         "GT",
         "GTE",
+        "HAS_PROPERTY",
+        "NOT_HAS_PROPERTY",
     ]
     value: HubSpotPropertyValue
 
 
-@purpose("Search contacts.")
+@dataclass
+class AndGroup:
+    """A group of conditions that must ALL be true (AND logic). Maximum of 6 conditions"""
+
+    conditions: List[HubSpotSearchCondition]
+
+
+@dataclass
+class SearchQuery:
+    """
+    Top-level search query that combines multiple AND groups with OR logic.
+    Any ONE of the AndGroups must match for a result to be returned.
+
+    Limitations:
+    - Maximum of 5 or_groups.
+    - Maximum 18 total conditions across all groups.
+    """
+
+    or_groups: List[AndGroup]
+
+
+def _convert_and_groups_to_filter_groups(
+    and_groups: List[AndGroup], schema: _HubSpotPropertiesSchema
+) -> List[dict[str, Any]]:
+    """
+    Convert our internal AND groups structure to HubSpot's filter groups format.
+    Each AND group becomes a filter group where all conditions must match.
+    """
+    filter_groups: List[dict[str, Any]] = []
+
+    for and_group in and_groups:
+        filters = []
+        for condition in and_group.conditions:
+            filter = {
+                "propertyName": condition.property_name,
+                "operator": condition.operator,
+            }
+            if condition.operator not in ("HAS_PROPERTY", "NOT_HAS_PROPERTY"):
+                value = _coerce_value_to_hubspot(
+                    name=condition.property_name,
+                    value=condition.value.value,
+                    schema=schema,
+                )
+                filter["value"] = value
+            filters.append(filter)
+
+        filter_groups.append({"filters": filters})
+    return filter_groups
+
+
 async def hubspot_search_contacts(
-    and_conditions: List[HubSpotSearchCondition],
-    created_after: Optional[datetime] = None,
-    created_before: Optional[datetime] = None,
+    query: SearchQuery,
     pagination_token: Optional[HubSpotPaginationToken] = None,
 ) -> Tuple[List[HubSpotContact], Optional[HubSpotPaginationToken]]:
-    """Search for HubSpot contacts
-    created_after: Return contacts that were created after this datetime
-    created_before: Return contacts that were created before this datetime
+    """Search for HubSpot contacts using OR-of-ANDs boolean logic.
+
+    The query follows OR-of-ANDs form: (A AND B AND C) OR (D AND E AND F) OR ...
+    where A, B, C, D, E, F are individual search conditions.
+
+    IMPORTANT: Each individual comparison should be in its own AND group when you want
+    to find matches for any of those comparisons (OR logic). Do NOT put multiple
+    OR conditions in the same AND group.
```

Comment:
I don't fully understand this last statement, when / how would it try to do this?

Reply:
Prior to adding this prompt, when fetching objects using multiple IDs, there was a tendency for it to do this:
```
# Process contacts in chunks of 6 (maximum AND conditions per OR group)
for i in range(0, len(contact_ids), 6):
    chunk_conditions: List[HubSpotSearchCondition] = []
    # Add conditions for each contact ID in this chunk
    for contact_id in contact_ids[i:i+6]:
        chunk_conditions.append(
            HubSpotSearchCondition(
                property_name="hs_object_id",
                operator="EQ",
                value=HubSpotPropertyValue(value=contact_id)
            )
        )
    search_groups.append(OrGroup(and_conditions=chunk_conditions))
```